### PR TITLE
Enable hot reloading of interface files

### PR DESCRIPTION
**Pull request by amylizzle**
[Screencast from 2024-05-12 14-03-37.webm](https://github.com/OpenDreamProject/OpenDream/assets/3855802/d61cefc3-241a-4871-b904-463e9d656fd0)

Adds a new proc to `/world` `ODHotReloadInterface` which forces a reload of the cached interface file and notifies all connected clients to reload it. 

Note this also cleans up some stuff that would probably causes issues with reconnecting too, so that's a nice bonus.

### CODE REVIEW

File: OpenDreamClient/Interface/DreamInterfaceManager.cs [hunk 1]
```diff
@@ -23,6 +23,7 @@
 using Robust.Shared.Utility;
 using SixLabors.ImageSharp;
 using System.Linq;
+using OpenToolkit.GraphicsLibraryFramework;
```
Comment (amylizzle) [inline OpenDreamClient/Interface/DreamInterfaceManager.cs:26]:
```suggestion
```

File: TestGame/code.dm [hunk 2]
```diff
@@ -193,6 +193,12 @@
 			src << "showing main window"
 			winset(src,"mainwindow","is-visible=true")
 
+	verb/test_hot_reload()
+		set category = "Test"
+		src << "tyring hot reload of interface..."
```
Comment (amylizzle) [inline TestGame/code.dm:198]:
```suggestion
		src << "trying hot reload of interface..."
```

**Review by amylizzle — COMMENTED** [review]
typo + using

File: OpenDreamRuntime/Procs/Native/DreamProcNativeWorld.cs [hunk 3]
```diff
@@ -185,6 +185,14 @@ internal static class DreamProcNativeWorld {
             return DreamValue.Null;
         }
 
+
```
Comment (amylizzle) [inline OpenDreamRuntime/Procs/Native/DreamProcNativeWorld.cs:188]:
```suggestion
```

File: OpenDreamRuntime/Resources/DreamResourceManager.cs [hunk 4]
```diff
@@ -78,9 +75,20 @@ public sealed class DreamResourceManager {
                         resource = new DreamResource(resourceId, resourcePath, resourcePath);
                         break;
                 }
+                return resource;
```
Comment (amylizzle) [inline OpenDreamRuntime/Resources/DreamResourceManager.cs:78]:
```suggestion

                return resource;
```

**Comment by wixoaGit** [comment]
I think it would be better to make this a server-side RT command

File: OpenDreamClient/Interface/DreamInterfaceManager.cs [hunk 5]
```diff
@@ -792,11 +794,20 @@ internal sealed class DreamInterfaceManager : IDreamInterfaceManager {
 
     private void Reset() {
         _userInterfaceManager.MainViewport.Visible = false;
-
+        //close windows if they're open, and clear all child uielements
+        foreach (var window in Windows.Values){
+            window.CloseChildWindow();
+            window.UIElement.RemoveAllChildren();
+        }
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/DreamInterfaceManager.cs:801]:
```suggestion
        }
        
```

File: OpenDreamClient/Interface/DreamInterfaceManager.cs [hunk 6]
```diff
@@ -792,11 +794,20 @@ internal sealed class DreamInterfaceManager : IDreamInterfaceManager {
...
         Windows.Clear();
         Menus.Clear();
         MacroSets.Clear();
+        //close popups if they're open
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/DreamInterfaceManager.cs:805]:
```suggestion
        
        //close popups if they're open
```

File: OpenDreamClient/Interface/DreamInterfaceManager.cs [hunk 7]
```diff
@@ -792,11 +794,20 @@ internal sealed class DreamInterfaceManager : IDreamInterfaceManager {
...
+        foreach (var popup in _popupWindows.Values) {
+            popup.Close();
+        }
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/DreamInterfaceManager.cs:808]:
```suggestion
        }
        
```

File: OpenDreamRuntime/DreamManager.Connections.cs [hunk 8]
```diff
@@ -278,5 +279,18 @@ public sealed partial class DreamManager {
         public DreamConnection GetConnectionBySession(ICommonSession session) {
             return _connections[session.UserId];
         }
+
+        public void HotReloadInterface() {
+            string? interfaceText = null;
+            if (_compiledJson.Interface != null)
+                interfaceText = _dreamResourceManager.LoadResource(_compiledJson.Interface, forceReload:true).ReadAsString();
+            var msgLoadInterface = new MsgLoadInterface() {
```
Comment (wixoaGit) [inline OpenDreamRuntime/DreamManager.Connections.cs:287]:
```suggestion

            var msgLoadInterface = new MsgLoadInterface() {
```

File: OpenDreamRuntime/DreamManager.Connections.cs [hunk 9]
```diff
@@ -278,5 +279,18 @@ public sealed partial class DreamManager {
...
+                InterfaceText = interfaceText
+            };
+            foreach (var connection in _connections.Values) {
```
Comment (wixoaGit) [inline OpenDreamRuntime/DreamManager.Connections.cs:290]:
```suggestion

            foreach (var connection in _connections.Values) {
```

File: OpenDreamRuntime/DreamManager.Connections.cs [hunk 10]
```diff
@@ -278,5 +279,18 @@ public sealed partial class DreamManager {
...
+                connection.Session?.Channel.SendMessage(msgLoadInterface);
+            }
+        }
     }
+
```
Comment (wixoaGit) [inline OpenDreamRuntime/DreamManager.Connections.cs:295]:
```suggestion
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]
Linter seems to have missed these whitespace issues.

**Comment by amylizzle** [comment]
> I think it would be better to make this a server-side RT command

I think it's actually better to leave this as a DM side action, so it can be easily integrated into existing admin and debug tooling which is pretty much all DM already.

The eventual plan here is to integrate it with hot reloading actions of IDEs, but until that's in place I think this is the best approach.

**Comment by wixoaGit** [comment]
> > I think it would be better to make this a server-side RT command
> 
> I think it's actually better to leave this as a DM side action, so it can be easily integrated into existing admin and debug tooling which is pretty much all DM already.
> 
> The eventual plan here is to integrate it with hot reloading actions of IDEs, but until that's in place I think this is the best approach.

Can we get an RT command regardless so this can be used on codebases that haven’t added it?

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**